### PR TITLE
Add Fourier transform playground example

### PR DESCRIPTION
**Pull request by brisvag**
# Description

Adds a new example to play around with 2D sine waves and fourier transforms. I made it to better visualise and explain how power spectrum and phase behave depending on the fourier components.

The example is also a nice showcase of two-way threading communication :P

Here's what it looks like:

https://github.com/napari/napari/assets/23482191/522f3e8d-cbe5-41c4-9c78-9da05d751adc

### CODE REVIEW

**Comment by stefanv** [comment]
Liking this very much!!

FFT visualization often benefits from taking the log, to avoid large changes in amplitude. Something like `np.log10(c + abs(spectrum))` and then playing with the value of `c`, given the application.

File: examples/fourier_transform_playground.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+FPS = 20
+
+# meshgrid used to calculate the 2D sine waves
+x = np.arange(IMAGE_SIZE) - IMAGE_SIZE / 2
+X, Y = np.meshgrid(x, x)
+
+
+def wave_2d(wavelength, angle, phase_shift, speed):
+    """
+    Generate a 2D sine wave based on angle and wavelength.
```
Comment (jni) [inline examples/fourier_transform_playground.py]:
```suggestion
    """Generate a 2D sine wave based on angle and wavelength.
```

File: examples/fourier_transform_playground.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+
+# set up viewer with grid-mode enabled
+viewer = napari.Viewer()
+viewer.grid.enabled = True
+
+
+def update_layer(name, data, **kwargs):
+    """
+    Update a layer in the viewer with new data.
```
Comment (jni) [inline examples/fourier_transform_playground.py]:
```suggestion
    """Update a layer in the viewer with new data.
```

File: examples/fourier_transform_playground.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        viewer.add_image(data, name=name, interpolation2d='spline36', **kwargs)
+        viewer.reset_view()
+    else:
+        viewer.layers[name].data = data
+
+
+def combine_and_set_data(waves):
+    """
+    Merge 2D waves, calculate the FT and update the viewer.
+    """
```
Comment (jni) [inline examples/fourier_transform_playground.py]:
```suggestion
    """Merge 2D waves, calculate the FT and update the viewer."""
```

File: examples/fourier_transform_playground.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+def moving_wave(
+    wave_id: int = 0,
+    wavelength: int = IMAGE_SIZE // 2,
+    angle: int = 0,
+    phase_shift: int = 0,
+    speed: float = 1,
+    run=True,
+):
+    """
+    Send new parameters to the listening thread to update the 2D waves.
```
Comment (jni) [inline examples/fourier_transform_playground.py]:
```suggestion
    """Send new parameters to the listening thread to update the 2D waves.
```

**Review by jni — APPROVED** [review]
I have to say I was *very* confused about where the animation was coming from... Then I saw the sneaky `time()` call way inside `wave_2d`! That'll learn me to read docstrings 😂 *Personally*, I would prefer for `wave_2d` to be purely functional, take in the current time, and the higher-level loop (`update_viewer`) to pass in the time. But I wouldn't hold up the PR for it.

At any rate, I played with it, and it's so mesmerising.... I love it. 😍 😵‍💫 😃

**Comment by brisvag** [comment]
> FFT visualization often benefits from taking the log, to avoid large changes in amplitude. Something like np.log10(c + abs(spectrum)) and then playing with the value of c, given the application.

Nice! I settled on `c = 10` for this application, it improves the readability of the phase.

Speaking of which, I fixed the phase, since I was doing it wrong before: I was taking only the imaginary part instead of getting the angle of the complex number.

I also moved out the time component from the base function, which I agree makes things cleaner. Should be good to go!

**Comment by GenevieveBuckley** [comment]
This is great!

It reminds me very strongly of Grant Jensen's [image Fourier transform talk](https://www.youtube.com/watch?v=nyk75ufbP74&list=PLhiuGaXlZZenm7lu5qv_A59zEWkRKkBn5&index=17), which is part of his [incredible Cryo-EM lecture series](https://www.youtube.com/playlist?list=PLhiuGaXlZZenm7lu5qv_A59zEWkRKkBn5) (seriously you should go look, it's brilliant and there are several other fourier transform videos in that series as well as the one I linked above)

Can we add links to those in the text description as well? I think that would really help boost people's understanding.

**Comment by jni** [comment]
The docs build is failing, I presume because some of the threading / interactive examples need to be excluded from running in the gallery. (But tbh I don't know, the fail log doesn't offer much of a clue...) I can't remember how this works... @melissawm any ideas? 😬 🙏

**Comment by brisvag** [comment]
> It reminds me very strongly of Grant Jensen's [image Fourier transform talk](https://www.youtube.com/watch?v=nyk75ufbP74&list=PLhiuGaXlZZenm7lu5qv_A59zEWkRKkBn5&index=17), which is part of his [incredible Cryo-EM lecture series](https://www.youtube.com/playlist?list=PLhiuGaXlZZenm7lu5qv_A59zEWkRKkBn5)

Yeah, that's part of the go-to lecture series we always suggest to newbies (me included :P) here :) Being able to interact with it is so much better though. Explaining aliasing by moving a slider and seing the peaks in the power spectrum "bounce" on the edges is so satisfying :)

**Comment by melissawm** [comment]
I tested locally and couldn't find any issues (docs build just fine), but I restarted the CI job yesterday and was watching - it seems like the examples grow progressively slower after this new one, maybe something is leaking? Not sure how to debug locally as I don't see any issues but will try to poke around. Sorry I'm not too familiar with the threading stuff!

**Comment by brisvag** [comment]
Adding `thread.quit()` seems to solve the docs bulding issue, but now there's some odd segfault in pyside6, which also seem to refer to threads... may be related?

File: examples/fourier_transform_playground.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        thread.send((wave_id, frequency, angle, phase_shift, speed))
+
+
+wdg = moving_wave()
+
+# add the widget to the window and run it once
+viewer.window.add_dock_widget(wdg, area='bottom')
+wdg()
+
+napari.run()
```
Comment (jni) [inline examples/fourier_transform_playground.py:149]: I think this should be protected by `if __name__ == '__main__':`, too? I'm a bit confused why it even works LOL

File: examples/fourier_transform_playground.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+thread.quit()
```
Comment (jni) [inline examples/fourier_transform_playground.py:151]: nice catch 😅

File: examples/fourier_transform_playground.py (hunk 5 above)
Comment (brisvag) [reply examples/fourier_transform_playground.py:149]: The issue is that to define some of the callbacks, the thread needs to already be spawned... so I would have to stick a lot of stuff inside the `if __name__` block... not sure that helps much :P

**Pull request merged**